File: stack.go
package main

import (
	"fmt"
)
const defaultSize int = 1
type Stack struct {
	elements []int
	top int
	capacity int
}
func NewStack() *Stack {
	return &Stack{ 
		elements: make([]int, defaultSize),
		top: 0,
		capacity: defaultSize,
	}
}
func (s *Stack) Push(value int) {
	if s.top == s.capacity {
		s.resize()
	}
	s.elements[s.top] = value
	s.top++
}
func (s *Stack) resize(){
	newCapacity := s.capacity * 2
	if newCapacity == 0 {
		newCapacity = 1
	}

	newElements := make([]int, newCapacity)
	for i:=0; i <s.top ; i++ {
		newElements[i] = s.elements[i]
	}
	s.elements = newElements
	s.capacity = newCapacity
	fmt.Printf("-- Memory resized %d --> %d \n" , s.capacity/2 , s.capacity)
}

func (s *Stack) Pop() (int, bool) {
	if s.top == 0 {
		return 0, false
	}
	s.top--;
	value := s.elements[s.top]
	return value, true
}

func (s *Stack) isEmpty() bool {
	return s.top == 0;
}

func main() {
	s := NewStack()
	s.Push(2)
	s.Push(5)
	if value, ok := s.Pop(); ok {
		fmt.Printf("%d pop\n", value)
	}
	
}

File: stack.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;
class Stack
{
private:
    vector<int> elements;     
public:
    void push(int const& element)
    {
        elements.push_back(element);
    }
    void pop()
    {
        if (elements.empty())
        {
            throw out_of_range("Stack<>::pop(): empty stack");
        }
        elements.pop_back();
    }
    int top() const
    {
        if (elements.empty())
        {
            throw out_of_range("Stack<>::top(): empty stack");
        }
        return elements.back();
    }
    bool isEmpty() const
    {
        return elements.empty();
    }
    size_t size() const
    {
        return elements.size();
    }
};


int main()
{
    vector<string> msg {"Hello", "C++", "World", "from", "VS Code", "and the C++ extension!"};
    Stack stack;
    for (const string& word : msg)
    {
        stack.push(word.length());
    }
    cout << "The lengths of the words are: ";   

    for (size_t i = 0; i < msg.size(); ++i)
    {
        cout << stack.top() << " ";
        stack.pop();
    }
    cout << endl;
}
